week4/week4.playground/Contents.swift

```diff
@@ -0,0 +1,203 @@
+// Hint: Do not use if statements to complete this section.
+// Call the function twice, sending a different Animal each time.
+
+enum Animal: String {
+  case dog, cat, lion, tiger, fox
+}
+
+func theSoundMadeBy(animal: Animal) -> String {
+  switch animal {
+  case .dog:
+    return "A dog goes woof"
+  case .cat:
+    return "A cat goes meow"
+  case .lion:
+    return "A lion roar"
+  case .tiger:
+    return "A tiger growl"
+  case .fox:
+    return "A fox screams"
+    
+  }
+}
+
+let catSound = theSoundMadeBy(animal: .cat)
+let dogSound = theSoundMadeBy(animal: .dog)
+print(catSound)
+print(dogSound)
+
+
+/*c) This question will have you creating multiple functions that will require you to use closures and collections. First, you will do some setup.
+
+Create an array of Int called nums with the values of 0 to 100.
+Create an array of Int? called numsWithNil with the following values:
+79, nil, 80, nil, 90, nil, 100, 72
+
+Create an array of Int called numsBy2 with values starting at 2 through 100, by 2.
+Create an array of Int called numsBy4 with values starting at 2 through 100, by 4.
+
+You can set the values of the arrays above using whatever method you find the easiest. In previous weeks you were introduced to ranges and sequences in Swift. Leveraging those in the Array initializer will allow you to create the requested arrays in a single line. Don’t let the last two break your stride.
+
+- Create a function called evenNumbersArray that takes a parameter of [Int] (array of Int) and returns [Int]. The array of Int returned should contain all the even numbers in the array passed. Call the function passing the nums array and print the output.
+
+- Create a function called sumOfArray that takes a parameter of [Int?] and returns an Int. The function should return the sum of the array values passed that are not nil. Call the function passing the numsWithNil array, and print out the results.
+
+- Create a function called commonElementsSet that takes two parameters of [Int] and returns a Set<Int> (set of Int). The function will return a Set<Int> of the values in both arrays.
+
+- Call the function commonElementsSet passing the arrays numsBy2, numsBy4, and print out the results.
+*/
+
+let nums = Array(0...100)
+let numsWithNil: [Int?] = [79, nil, 80, nil, 90, nil, 100, 72]
+
+var numsBy2 = [Int]()
+for number in stride(from: 2, to: 100, by: 2) {
+  numsBy2.append(number)
+}
+print("numsBy2 array will be \(numsBy2)")
+
+var numsBy4 = [Int]()
+for number in stride(from: 2, to: 100, by: 4) {
+  numsBy4.append(number)
+}
+print("numsBy4 array will be \(numsBy4)")
+
+func evenNumbersArray(numberArray: [Int]) -> [Int] {
+  var evenNumbers = numberArray.filter { $0 % 2 == 0 }
+  return evenNumbers
+}
+
+print(evenNumbersArray(numberArray: nums))
+
+func sumOfArray(optionalArray: [Int?]) -> Int {
+  var sumOfNonNilValues = optionalArray.compactMap { Int($0 ?? 0)}.reduce(0, +)
+  return sumOfNonNilValues
```

Comment:
Same thing as above here. But also, additionally, No need to cast it into an `Int` and provide a default value. compactMap serves exactly for that, to remove any nil values. So this would be the correct solution: 
`optionalArray.compactMap { $0 }.reduce(0, +)`